server/src-lib/Hasura/RQL/DDL/Schema/Table.hs

```diff
@@ -79,33 +85,80 @@ $(deriveJSON (aesonDrop 2 snakeCase){omitNothingFields=True} ''UntrackTable)
+trackExistingTableOrViewP1 qt = do
   adminOnly
   rawSchemaCache <- askSchemaCache
-  when (M.member vn $ scTables rawSchemaCache) $
-    throw400 AlreadyTracked $ "view/table already tracked : " <>> vn
-  let qf = fmap (FunctionName . getTableTxt) vn
+  when (M.member qt $ scTables rawSchemaCache) $
+    throw400 AlreadyTracked $ "view/table already tracked : " <>> qt
+  let qf = fmap (FunctionName . getTableTxt) qt
   when (M.member qf $ scFunctions rawSchemaCache) $
-    throw400 NotSupported $ "function with name " <> vn <<> " already exists"
+    throw400 NotSupported $ "function with name " <> qt <<> " already exists"
+
+validateTableConfig
+  :: (QErrM m, CacheRM m)
+  => TableInfo a -> TableConfig -> m ()
+validateTableConfig tableInfo (TableConfig rootFlds colFlds) =
+  withPathK "configuration" $ do
+    withPathK "custom_root_fields" $ do
+      sc <- askSchemaCache
+      let defRemoteGCtx = scDefaultRemoteGCtx sc
+      GS.validateCustomRootFlds defRemoteGCtx rootFlds
+    withPathK "custom_column_fields" $
+      forM_ (M.toList colFlds) $ \(col, GraphQLName customName) -> do
+        void $ askPGColInfo (_tiFieldInfoMap tableInfo) col ""
+        withPathK (getPGColTxt col) $
+          checkForFldConfilct tableInfo $ FieldName $ G.unName customName
+        when (not $ null duplicateNames) $ throw400 NotSupported $
+          "the following names are duplicated: " <> showNames duplicateNames
+  where
+    duplicateNames = duplicates $ map unGraphQLName $ M.elems colFlds
 
 trackExistingTableOrViewP2
   :: (QErrM m, CacheRWM m, MonadTx m, MonadIO m, HasHttpManager m, HasSQLGenCtx m)
-  => TrackTable -> m EncJSON
-trackExistingTableOrViewP2 (TrackTable tableName isEnum) = do
+  => QualifiedTable -> Bool -> TableConfig -> m EncJSON
+trackExistingTableOrViewP2 tableName isEnum config = do
   sc <- askSchemaCache
   let defGCtx = scDefaultRemoteGCtx sc
   GS.checkConflictingNode defGCtx $ GS.qualObjectToName tableName
-  saveTableToCatalog tableName isEnum
+  saveTableToCatalog tableName isEnum config
   buildSchemaCacheFor (MOTable tableName)
   return successMsg
 
 runTrackTableQ
   :: (QErrM m, CacheRWM m, MonadTx m, UserInfoM m, MonadIO m, HasHttpManager m, HasSQLGenCtx m)
   => TrackTable -> m EncJSON
-runTrackTableQ q = do
-  trackExistingTableOrViewP1 q
-  trackExistingTableOrViewP2 q
+runTrackTableQ (TrackTable qt isEnum) = do
+  trackExistingTableOrViewP1 qt
+  trackExistingTableOrViewP2 qt isEnum noTableConfig
+
+data TrackTableV2
+  = TrackTableV2
+  { ttv2Table         :: !TrackTable
+  , ttv2Configuration :: !TableConfig
+  } deriving (Show, Eq, Lift)
+$(deriveJSON (aesonDrop 4 snakeCase) ''TrackTableV2)
+
+runTrackTableV2Q
+  :: (QErrM m, CacheRWM m, MonadTx m, UserInfoM m, MonadIO m, HasHttpManager m, HasSQLGenCtx m)
+  => TrackTableV2 -> m EncJSON
+runTrackTableV2Q (TrackTableV2 (TrackTable qt isEnum) config) = do
+  sc <- askSchemaCache
+  let tableCache = scTables sc
+  case (M.lookup qt tableCache) of
```

Comment:
I'm planning to implement updating **only** table configuration on version `2` `track_table` query if the table is tracked already. The `isEnum` state remains untouched.

Reply:
That seems kind of inconsistent, though, isn’t it? Shouldn’t we either update them both or update neither? I just think these kinds of APIs are confusing, since as a user it (1) isn’t clear what the behavior is and (2) is dangerous to write code that does a `PUT` because a future update to the API could add a new field that your old code will unintentionally clobber.

Whatever we decide, `POST` is definitely not the right HTTP verb for this if we’re trying to be at all RESTful—it should be `PUT` or `PATCH`. That said, I don’t actually care all that much about being RESTful, so long as we have our own set of API design principles.

Reply:
I agree. The solution I'm seeing now is to have a separate query type in version `2` to update the configuration, like `update_table_configuration`, which is analogous to `set_table_is_enum` API. Is that okay?

Reply:
Sure. I'm going with `set_table_custom_fields`.